File: tlg_android/src/ch/bailu/tlg_android/GameActivity.java
package ch.bailu.tlg_android;


import java.io.BufferedOutputStream;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

public class GameActivity extends Activity implements Runnable, SurfaceHolder.Callback {
    
    private Handler               timer=null;
    private GameView   tetris;
    private MotionEventTranslater motionTranslater;
    
    
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        SurfaceView surface = new SurfaceView(this);
        surface.getHolder().addCallback(this);
        setContentView(surface);
        
        motionTranslater=new MotionEventTranslater();
   }

 
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        inflateMenu(R.menu.menu, menu);
        return true;
    }
    
    protected void inflateMenu(int id, Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(id, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        
        switch (item.getItemId()) {
        case R.id.menu_grid:
            tetris.toggleGrid();
            return true;
        case R.id.menu_pause:
            tetris.pauseOrResume();
            startTimer();
            return true;
        case R.id.menu_start:
            tetris.newGame();
            startTimer();
            return true;
        case R.id.menu_about:
        	start(AboutActivity.class);
        	return true;
        case R.id.menu_score:
        	start(HighscoreActivity.class);
        	return true;
        case R.id.menu_name:
            tetris.setHighscoreName();
        default:
            return false;
        }
    }

    
    private void start(Class<?> activityClass) {
    	Intent intent = new Intent();
        intent.setClass(this, activityClass); 
        intent.setAction(activityClass.getName());
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);         
    }

    
    private void startTimer() {
        if (timer==null) {
            timer = new Handler();
            kickTimer();
        }
    }
    
    private void stopTimer() {
        if (timer != null) {
            timer.removeCallbacks(this);
            timer=null;
        }
    }
    
    private void kickTimer() {
        if (timer != null) {
            int interval = tetris.getTimerInterval();
            if (interval==0) {
                stopTimer();
            } else {
                timer.removeCallbacks(this);
                timer.postDelayed(this, interval); // fire in milliseconds
            }
        }
    }

    @Override
    public void run() {
        tetris.moveShape(MotionEventTranslater.KEY_DOWN);
        kickTimer();
    }
   
    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width,
            int height) {
        tetris.setPixelGeometry(width,height);
        motionTranslater.setGeometry(width,height);
    }


    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        tetris = new GameView(this);
        startTimer();
        tetris.startPainter(this,holder);
    }


    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        tetris.stopPainter();
        stopTimer();
        writeState();
    }
    
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (motionTranslater.translateEvent(event))
            tetris.moveShape(motionTranslater.getLastEvent());
        return true;
    }
 
    
    private final static String STATE_FILE = "app_state";
    
    private void writeState() {
        try {
            BufferedOutputStream output = new BufferedOutputStream( openFileOutput(STATE_FILE, Context.MODE_PRIVATE));
            tetris.writeState();
            output.close();
        } catch (Exception e) {
            // FIXME do something
        }
        
    }

}

File: tlg_android/src/main/java/ch/bailu/tlg_android/activity/GameActivity.kt
package ch.bailu.tlg_android.activity

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.view.*
import ch.bailu.tlg_android.R
import ch.bailu.tlg_android.controller.MotionEventTranslater
import ch.bailu.tlg_android.view.GameView
import java.io.BufferedOutputStream

class GameActivity : Activity(), Runnable, SurfaceHolder.Callback {
    private var timer: Handler? = null
    private var tetris: GameView? = null
    private var motionTranslater: MotionEventTranslater? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val surface = SurfaceView(this)
        surface.holder.addCallback(this)
        setContentView(surface)
        motionTranslater = MotionEventTranslater()
    }


    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        inflateMenu(R.menu.menu, menu)
        return true
    }

    private fun inflateMenu(id: Int, menu: Menu?) {
        val inflater: MenuInflater = getMenuInflater()
        inflater.inflate(id, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_grid -> {
                tetris!!.toggleGrid()
                true
            }
            R.id.menu_pause -> {
                tetris!!.pauseOrResume()
                startTimer()
                true
            }
            R.id.menu_start -> {
                tetris!!.newGame()
                startTimer()
                true
            }
            R.id.menu_about -> {
                start(AboutActivity::class.java)
                true
            }
            R.id.menu_score -> {
                start(HighscoreActivity::class.java)
                true
            }
            R.id.menu_name -> {
                tetris!!.setHighscoreName()
                false
            }
            else -> false
        }
    }


    private fun start(activityClass: Class<*>) {
        val intent = Intent()
        intent.setClass(this, activityClass)
        intent.action = activityClass.name
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        startActivity(intent)
    }


    private fun startTimer() {
        if (timer == null) {
            timer = Handler()
            kickTimer()
        }
    }

    private fun stopTimer() {
        if (timer != null) {
            timer!!.removeCallbacks(this)
            timer = null
        }
    }

    private fun kickTimer() {
        val timer = timer
        if (timer is Handler) {
            val interval = tetris!!.timerInterval
            if (interval == 0) {
                stopTimer()
            } else {
                timer.removeCallbacks(this)
                timer.postDelayed(this, interval.toLong()) // fire in milliseconds
            }
        }
    }

    override fun run() {
        tetris!!.moveShape(MotionEventTranslater.KEY_DOWN)
        kickTimer()
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        tetris!!.setPixelGeometry(width, height)
        motionTranslater!!.setGeometry(width, height)
    }


    override fun surfaceCreated(holder: SurfaceHolder) {
        tetris = GameView(this)
        startTimer()
        tetris!!.startPainter(this, holder)
    }


    override fun surfaceDestroyed(holder: SurfaceHolder) {
        tetris!!.stopPainter()
        stopTimer()
        writeState()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val motionTranslater = motionTranslater
        
        if (motionTranslater is MotionEventTranslater && motionTranslater.translateEvent(event)) tetris!!.moveShape(motionTranslater.getLastEvent())
        return true
    }

    private val STATE_FILE = "app_state"

    private fun writeState() {
        try {
            val output = BufferedOutputStream(openFileOutput(STATE_FILE, Context.MODE_PRIVATE))
            tetris!!.writeState()
            output.close()
        } catch (e: Exception) {
            System.err.println(e.message)
        }
    }
}
